pkg/storage/stores/shipper/downloads/table_manager.go

```diff
@@ -232,97 +240,107 @@ func (tm *TableManager) cleanupCache() error {
+		// continue if the table is not within query readiness
+		if activeTableNumber-tableNumber > int64(largestQueryReadinessNum) {
+			continue
+		}
+
+		// list the users that have dedicated index files for this table
+		_, usersWithIndex, err := tm.indexStorageClient.ListFiles(ctx, tableName)
 		if err != nil {
 			return err
 		}
 
-		tm.tablesMtx.Lock()
-		tm.tables[tableName] = table
-		tm.tablesMtx.Unlock()
+		// find the users whos index we need to keep ready for querying from this table
+		usersToBeQueryReadyFor := tm.findUsersInTableForQueryReadiness(tableNumber, usersWithIndex, queryReadinessNumByUserID)
 
-		err = table.EnsureQueryReadiness(context.Background())
+		// continue if both user index and common index is not required to be downloaded for query readiness
+		if len(usersToBeQueryReadyFor) == 0 && activeTableNumber-tableNumber > int64(tm.cfg.QueryReadyNumDays) {
+			continue
+		}
+
+		table, err := tm.getOrCreateTable(tableName)
 		if err != nil {
 			return err
 		}
+
+		if err := table.EnsureQueryReadiness(ctx, usersToBeQueryReadyFor); err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
-// queryReadyTableNumbersRange returns the table numbers range. Table numbers are added as suffix to table names.
-func (tm *TableManager) queryReadyTableNumbersRange() (int64, int64) {
-	newestTableNumber := getActiveTableNumber()
+// findUsersInTableForQueryReadiness returns the users that needs their index to be query ready based on the tableNumber and
+// query readiness number provided per user
+func (tm *TableManager) findUsersInTableForQueryReadiness(tableNumber int64, usersWithIndexInTable []string,
+	queryReadinessNumByUserID map[string]int) []string {
+	activeTableNumber := getActiveTableNumber()
+	usersToBeQueryReadyFor := []string{}
 
-	return newestTableNumber - int64(tm.cfg.QueryReadyNumDays), newestTableNumber
-}
-
-// tablesRequiredForQueryReadiness returns the names of tables required to be downloaded for being query ready as per configured QueryReadyNumDays.
-// It only considers daily tables for simplicity and we anyways have made it mandatory to have daily tables with boltdb-shipper.
-func (tm *TableManager) tablesRequiredForQueryReadiness(tablesInStorage []string) ([]string, error) {
-	// regex for finding daily tables which have a 5 digit number at the end.
-	re, err := regexp.Compile(`.+[0-9]{5}$`)
-	if err != nil {
-		return nil, err
-	}
-
-	minTableNumber, maxTableNumber := tm.queryReadyTableNumbersRange()
-	var requiredTableNames []string
-
-	for _, tableName := range tablesInStorage {
-		if !re.MatchString(tableName) {
-			continue
+	for _, userID := range usersWithIndexInTable {
+		// use the query readiness config for the user if it exists or use the default config
+		queryReadyNumDays, ok := queryReadinessNumByUserID[userID]
+		if !ok {
+			queryReadyNumDays = tm.cfg.Limits.DefaultLimits().QueryReadyIndexNumDays
```

Comment:
This is not consistent with above where you first check the config and after the limit.

I wonder if we should in fact remove the config ? Or deprecate it ?